### ISSUE TITLE
Singleton Guidance

### ISSUE THREAD

**Issue by shellscape** [markdown-it/markdown-it#815]
This is likely not the intended use for markdown-it, and if that's the answer, I will accept that. However, I was wondering if you have any thoughts on how to use an instance of markdown-it as a singleton. I'm working on a rather intense API and squeezing every bit of performance out of it. One of the areas I'm targeting is making everything that can be a singleton, a singleton. Here's the gist of how I have markdown-it setup:

```ts
import MarkdownIt from 'markdown-it';
// ... I'm importing a bunch of plugins. removed to save space ...
import toc, { TocAst } from 'markdown-it-toc-done-right';

const markdown = new MarkdownIt({
  html: true,
  linkify: true,
  typographer: true
});

// I'm using a standard set of plugins, which require no options, for each invocation of `markdown.render`
[ ... ].forEach((plugin) => markdown.use(plugin));

export const render = async (
  md: string,
  { useToc = true }
) => {
  const receiveToc = useToc
    ? new Promise<TocAst>((resolve) => {
        markdown.use(toc, {
          callback: (_: string, ast: TocAst) => resolve(ast)
        });
      })
    : Promise.resolve({} as any);

  const html = markdown.render(md);
  const tocAst = useToc ? await receiveToc : null;

  return { html, toc: normalizeToc(tocAst) } as MarkdownParserResult;
};
```

This may look well and good, but it has the unintended consequence of adding the `toc` plugin N+1 number of times for each invocation. As such, the `toc` plugin chokes. This is probably due to how the `toc` plugin manages some state between rules and such, but I'm not going to ask the developer there to rework their plugin for my edge case, especially when `markdown` _doesn't have to be_ a singleton.

All that said, is there an approach I'm not seeing here for using a "base" markdown-it instance that already has plugins added, and instantiate from that "base" instance for each invocation of `render`?

**Comment by puzrin** [comment]
> All that said, is there an approach I'm not seeing here for using a "base" markdown-it instance that already has plugins added, and instantiate from that "base" instance for each invocation of render?

IMO that's not possible.

- Each `markdown-it` instance should be configured from scratch. That defines which plugins/options will be used (for every document). You can have multiple singletones and select appropriate for every concrete case.
- Document state, if produced by plugins, should be stored in `env` param. It passes document local state between plugins and from parser to renderer.

In theory, you could pass your "document options" via `env` at start, and make your plugin skip on some flag https://github.com/markdown-it/markdown-it/blob/master/lib/index.js#L541. But IMO that's dirty. This was reserved for very advanced cases. For example, if you have multiple documents on single page with footnotes, and need to avoid link conflicts - you can pass link prefix to each document.

---

if you need async processing

- it can be done between parser/renderer call (if you invoke those manually, instead of `.render()`)
- it can be called after markdown (independent), over HTML.

https://github.com/markdown-it/markdown-it/blob/master/docs/development.md#faq

---

In short - recheck your data flow, and make sure every code unit keep data [state] in appropriate place (in `markdown-it` instance or in `env`). If your toc plugin do it different - fix it.
